Add To/From Proto Round Trip for ExecutableStage

tgroh opened the pull request:
Make ExecutableStage an abstract class with a well-defined (and final)
equals and hashCode methods.

Add an ImmutableExecutableStage which is built from the components of
the ExecutableStage.

------------------------

Follow this checklist to help us incorporate your contribution quickly and easily:

 - [x] Make sure there is a [JIRA issue](https://issues.apache.org/jira/projects/BEAM/issues/) filed for the change (usually before you start working on it).  Trivial changes like typos do not require a JIRA issue.  Your pull request should address just this issue, without pulling in other changes.
 - [x] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue.
 - [x] Write a pull request description that is detailed enough to understand:
   - [x] What the pull request does
   - [x] Why it does it
   - [x] How it does it
   - [x] Why this approach
 - [x] Each commit in the pull request should have a meaningful subject line and body.
 - [x] Run `mvn clean verify` to make sure basic checks pass. A more thorough check will be performed on your pull request automatically.
 - [x] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

tgroh commented:
R: @bsidhom

bsidhom commented on runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/graph/ExecutableStage.java:
```diff
@@ -80,5 +93,86 @@
    * #getOutputPCollections()}.
    */
-  PTransform toPTransform();
+  public final PTransform toPTransform() {
```
Why not keep this as an interface and make this a default method?  Targeting Java 8, the static method below can also live in the interface.

tgroh replied:
I want to force an overload of equals and hashCode, and got pushback from my compiler (I think). Certainly AutoValue clobbered me there.

bsidhom replied:
Since we no longer need `hashCode` or `equals`, I prefer an interface for aesthetic reasons. I'll leave the final judgement up to you between this and ensuring that `toPTransform` is final.

tgroh replied:
Done.

To be honest, this has many of the same restrictions as any `XPayload` object - it's got a known format, but all translators using the URN must produce the same format of stage, otherwise either preconditions will fail or code will be buggy.

Documented that proto shape, and moved to an interface.

tgroh commented:
Alright I have added significant changes to this PR

Specifically I have renamed `GreedilyFused....` to be a `GreedyStageFuser` factory for instances of `ExecutableStage`, as it's a property of construction not of the actual contents. 

I am keeping `ExecutableStage` as an abstract class rather than interface to force the `toPTransform` method to be exactly one implementation within Java, but I'm willing to move here if would be preferred to be kept an interface.

lukecwik commented:
I'll take a look once @bsidhom has taken a pass.

bsidhom commented on runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/Environments.java:
```diff
@@ -56,6 +57,27 @@
+        return Optional.empty();
+      }
+    } catch (IOException e) {
+      throw new RuntimeException(e);
```
It's a little strange to have an exception-throwing method that returns an optional. On the other hand, IOExceptions _really_ shouldn't happen here, so maybe this doesn't matter.

tgroh replied:
I don't really agree with the former claim, though I also agree strongly with the latter.

Optional is the `maybe` type - in the happy case, we always return a thing, even if that thing is the explicit absence of a thing. If we fail for any reason (we got a malformed proto, or whatever), there's not meaningfully a result - it's not unknown, it's a precondition failure (this could probably be `throw new IllegalArgumentException("complaints", e)`, for example).

bsidhom replied:
I agree in practice, but only because Java doesn't provide `either` for exceptional outcomes. In any case, it's irrelevant here. 😉

bsidhom commented on runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/graph/GreedyStageFuser.java:
```diff
@@ -0,0 +1,176 @@
+                "%s must be populated on all %s in a %s",
+                Environment.class.getSimpleName(),
+                PTransformNode.class.getSimpleName(),
+                GreedyStageFuser.class.getSimpleName()));
```
Don't construct the exception unconditionally. Since you're passing in a lambda anyway, you might just define a supplier here. Alternatively, refactor this into a separate method.

tgroh replied:
Done.

bsidhom commented on runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/graph/ImmutableExecutableStage.java:
```diff
@@ -0,0 +1,52 @@
+
+/** An {@link ExecutableStage} which is constructed with all of its initial state. */
+@AutoValue
+abstract class ImmutableExecutableStage extends ExecutableStage {
```
Does this really compile after reverting to `interface`?

tgroh replied:
Strangely enough, it does not :)

Fixed.

tgroh commented:
@lukecwik PTaL

lukecwik commented on runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/Environments.java:
```diff
@@ -56,6 +57,27 @@
+              .getOrDefault(ptransform.getSpec().getUrn(), DEFAULT_SPEC_EXTRACTOR)
+              .getEnvironmentId(ptransform);
+      if (!Strings.isNullOrEmpty(envId)) {
+        // Some PTransform payloads may have an empty (default) Environment ID, for example a
```
The default value in a proto string field is the empty field.

So its better to word this as an "unspecified" environment id.

tgroh replied:
Done.
